Detox 20 (#3285)
wix/Detox: docs/guide/test-id.mdx

```jsx title="YourScreen.jsx"
const ITEMS = [
  { title: 'First Item' },
  { title: 'Second Item' },
  { title: 'Third Item' },
];

function YourScreen() {
  const renderItem = ({ item }) => (
// highlight-next-line
    <YourCustomComponent testID={'listItem'} lable={item.title} />
  );

  return (
      <FlatList
        data={ITEMS}
        renderItem={renderItem}
      />
  );
}
```

This would be a violation of accessibility guidelines and unnecessary complication for your test matchers.
You’d also have to use extra matchers and `.atIndex` clarification:

```js
expect(element(by.id('listItem')).atIndex(2)).toHaveText('Third Item');
```

Instead, you could generate a unique `testID` for every list item with the `index` property:

```jsx
  const renderItem = ({ item, index }) => (
    <YourCustomComponent testID={`listItem.${index + 1}`} lable={item.title} />
  );
```

That way, your assertion would become simpler and more deterministic:

```js
expect(element(by.id('listItem.3'))).toHaveText('Third Item');
```

![testID for repetitive components](../img/repetitiveComponentTestID.png)

## Find your testID

:::note

> Incorrect or absent `testID` is a common cause for test failure.

Reviewer: ```suggestion
Having an incorrect or nonexistent `testID` is a common cause of test failure.
```
